solr/core/src/test/org/apache/solr/util/BaseTestCircuitBreaker.java

```diff
@@ -0,0 +1,314 @@
+                    MatcherAssert.assertThat(
+                        e.getMessage(), containsString("Circuit Breakers tripped"));
+                    failureCount.incrementAndGet();
+                  } catch (Exception e) {
+                    throw new RuntimeException(e.getMessage());
+                  }
+                });
+
+        futures.add(future);
+      }
+
+      for (Future<?> future : futures) {
+        try {
+          future.get();
+        } catch (Exception e) {
+          throw new RuntimeException(e.getMessage());
+        }
+      }
+    } finally {
+      ExecutorUtil.shutdownAndAwaitTermination(executor);
+      assertEquals("Number of failed queries is not correct", 5, failureCount.get());
+    }
+  }
+
+  public void testResponseWithCBTiming() {
+    removeAllExistingCircuitBreakers();
+
+    assertQ(
+        req("q", "*:*", CommonParams.DEBUG_QUERY, "true"),
+        "//str[@name='rawquerystring']='*:*'",
+        "//str[@name='querystring']='*:*'",
+        "//str[@name='parsedquery']='MatchAllDocsQuery(*:*)'",
+        "//str[@name='parsedquery_toString']='*:*'",
+        "count(//lst[@name='explain']/*)=3",
+        "//lst[@name='explain']/str[@name='1']",
+        "//lst[@name='explain']/str[@name='2']",
+        "//lst[@name='explain']/str[@name='3']",
+        "//str[@name='QParser']",
+        "count(//lst[@name='timing']/*)=3",
+        "//lst[@name='timing']/double[@name='time']",
+        "count(//lst[@name='prepare']/*)>0",
+        "//lst[@name='prepare']/double[@name='time']",
+        "count(//lst[@name='process']/*)>0",
+        "//lst[@name='process']/double[@name='time']");
+
+    CircuitBreaker circuitBreaker = new MockCircuitBreaker(false);
+    h.getCore().getCircuitBreakerRegistry().register(circuitBreaker);
+
+    assertQ(
+        req("q", "*:*", CommonParams.DEBUG_QUERY, "true"),
+        "//str[@name='rawquerystring']='*:*'",
+        "//str[@name='querystring']='*:*'",
+        "//str[@name='parsedquery']='MatchAllDocsQuery(*:*)'",
+        "//str[@name='parsedquery_toString']='*:*'",
+        "count(//lst[@name='explain']/*)=3",
+        "//lst[@name='explain']/str[@name='1']",
+        "//lst[@name='explain']/str[@name='2']",
+        "//lst[@name='explain']/str[@name='3']",
+        "//str[@name='QParser']",
+        "count(//lst[@name='timing']/*)=4",
+        "//lst[@name='timing']/double[@name='time']",
+        "count(//lst[@name='circuitbreaker']/*)>0",
+        "//lst[@name='circuitbreaker']/double[@name='time']",
+        "count(//lst[@name='prepare']/*)>0",
+        "//lst[@name='prepare']/double[@name='time']",
+        "count(//lst[@name='process']/*)>0",
+        "//lst[@name='process']/double[@name='time']");
+  }
+
+  private void removeAllExistingCircuitBreakers() {
+    List<CircuitBreaker> registeredCircuitBreakers =
+        h.getCore().getCircuitBreakerRegistry().getRegisteredCircuitBreakers();
+
+    registeredCircuitBreakers.clear();
```

Comment:
Could `deregisterAll` be used here instead (and then `getRegisteredCircuitBreakers()` could be removed)?

```suggestion
    h.getCore().getCircuitBreakerRegistry().deregisterAll();
```

Reply:
@cpoerschke Seems we had a race condition with your latest suggestions and my merge. Sorry for that, I could swear I saw some of those comments earlier, don't know how they got closed without merge.

I'll backport the initial commit. Would you mind filing a new PR with these latest changes? The commit message can reference same JIRA.

Reply:
No problem! #1870 for the follow-up bits. Would you mind including it in the back port?

Reply:
That was my thought, probably easier to correlate later..